packages/apollo-language-server/src/config/loadConfig.ts

```diff
@@ -50,26 +52,69 @@ export interface LoadConfigSettings {
 
   // used when run by a `Workspace` where we _know_ a config file should be present.
   requireConfig?: boolean;
-
-  // for CLI usage, we don't _require_ a config file for everything. This allows us to pass in
-  // options to build one at runtime
-  name?: string;
-  type?: "service" | "client";
 }
 
-export type ConfigResult<T> = {
+type NotNullConfigResult<T> = {
   config: T;
   filepath: string;
-} | null;
+};
+
+export type ConfigResult<T> = NotNullConfigResult<T> | null;
+
+export function loadConfigWithDefaults(
+  loadedConfig: Pick<
+    NotNullConfigResult<DeepPartial<ApolloConfigFormat>>,
+    "config"
+  > | null,
+  graphInfo: GraphInfo,
+  configPath?: string,
+  apiKey?: string
+): ApolloConfig {
+  const graphId = graphInfo.graphId;
+
+  const configWithDefaults = {
+    ...(loadedConfig && loadedConfig.config),
+    client: merge(
+      DefaultConfigBase,
+      DefaultClientConfig,
+      graphId
+        ? {
+            service: `${graphId}@${graphInfo.clientGraphVariant}`,
+            graphId
+          }
+        : {},
+      loadedConfig && loadedConfig.config.client
+    ),
+    service: merge(
+      DefaultConfigBase,
```

Comment:
We should put an `Object.create(null)` here in the first position to avoid mutation of `DefaultConfigBase`.